README.md
# XboxHunter
Программа для отслеживания наличия консолей XboxSeries S/X в магазинах.

## Для работы программы необходимо:
1) Установить [Python](https://www.python.org/downloads/) не ниже версии 3.6
2) Установить дополнительные пакеты:
   - requests
   - requests-html
   - beautifulsoup4
   - certifi
3) Запустить программу через файл [xbox_hunter.py](xbox_hunter.py) (требуется установленный pylauncher)

Подробное описание установки и настройки программы приведено ниже.

## Принцип работы
Программа каждые 1-3 минуты проверяет на веб-странице игровой консоли наличие/отсутствие следующих элементов:
- Ситилинк: кнопка "В корзину"
- Эльдорадо: кнопка "Сообщить о поступлении"
- Ozon: кнопка "Добавить в корзину"
- Связной: кнопка "Купить"
- Сбермегамаркет: кнопка "Купить"
- Gamepark: кнопки "Узнать о наличии" и "Только в магазине"
- Technopark: кнопка "Купить"
- Подпишись.РФ: кнопка "Оформить заявку"
- 1С-Интерес: кнопка "В корзину"
- КорпорацияЦентр: кнопка "Добавить в корзину"
- ОнлайнТрейд: статус "Снят с продажи"

В процессе работы программа выводит в своем окне время и результаты запроса.

При наличии игровой консоли программа открывает "удачную" ссылку в браузере, подает звуковой сигнал и (опционально) отправляет уведомление со ссылкой на электронную почту.

## Установка Python
1) Скачайте [Python](https://www.python.org/downloads/) не ниже версии 3.6
2) Запустите установочный файл, поставьте флажок напротив Add Python <номер версии> to PATH и нажмите Install Now. Запомните папку, в которую будет установлен Python
3) Скопируйте файл [install_packages.bat](install_packages.bat) в папку с файлом pip.exe. Например: C:\Program Files (x86)\Python3\Scripts\pip.exe
4) Запустите файл [install_packages.bat](install_packages.bat) для установки дополнительных пакетов

Теперь на вашем компьютере установлен Python с пакетами, необходимыми для работы программы.

## Настройка электронной почты для отправки уведомлений
1) Создайте новый Google-аккаунт (или используйте имеющийся аккаунт, если готовы снизить уровень его безопасности)
2) НЕ подключайте двухфакторную аутентификацию!
3) Перейдите в раздел Безопасность -> [Ненадежные приложения, у которых есть доступ к аккаунту](https://myaccount.google.com/lesssecureapps)
4) Включите опцию "Небезопасные приложения разрешены"

Периодически проверяйте настройки безопасности Google-аккаунта, т.к. опция "Небезопасные приложения разрешены" со временем автоматически отключается.

Без этой опции отправка уведомлений на электронную почту невозможна.

## Установка и настройка программы
1) Распакуйте архив с программой
2) Запустите файл [xbox_hunter.py](xbox_hunter.py)
3) Выберите необходимую игровую консоль для отслеживания
4) (Опционально) Введите логин и пароль Google-аккаунта, с которого будут отправляться уведомления (настройка аккаунта описана выше). Введите адрес электронной почты, на которую будут приходить уведомления

Учтите, что при вводе пароля символы не отображаются в окне программы. Подобная мера безопасности используется, например, в Ubuntu при работе с терминалом.

При первом запуске программы будет загружен Chromium, что может занять некоторое время. При последующих запусках программы загрузка Chromium не требуется.

## Настройка списка магазинов
При каждом запуске программа однократно считывает ссылки для отслеживания из текстовых файлов, расположенных в папках [xss_links](xss_links) и [xsx_links](xsx_links).

Для удаления магазина из списка отслеживания измените расширение файла с .txt на любое другое или удалите расширение файла.

Для возврата магазина в список отслеживания восстановите расширение файла .txt.

Пример:
- Ozon_xsx.txt - магазин в списке отслеживания
- Ozon_xsx - магазин не отслеживается

Из текстового файла считывается только первая строка, поэтому для добавления новых ссылок необходимо создать новый текстовый файл.

Все изменения в папках [xss_links](xss_links), [xsx_links](xsx_links) и текстовых файлах производите при закрытой программе.

## Если не запускается файл xbox_hunter.py
1) Откройте в текстовом редакторе файл [xbox_hunter.bat](xbox_hunter.bat)
2) В строке внутри первых кавычек укажите путь к файлу python.exe на вашем компьютере. Например: "C:\Program Files (x86)\Python3\python.exe" "xbox_hunter.py"
3) Запустите файл [xbox_hunter.bat](xbox_hunter.bat)

## Технические ограничения
1) На данный момент не реализовано отслеживание игровых консолей в магазинах:
   - МВидео
   - DNS
   - Мегафон
   - Computeruniverse
2) Программа не сообщает сайту магазина местоположение/город

## Примечания
- Программа разрабатывалась и тестировалась на версии [Python 3.8.10](https://www.python.org/ftp/python/3.8.10/python-3.8.10.exe)
- Для проверки звукового сигнала запустите файл [sound_test.py](test/sound_test.py) (находится в папке [test](test))


CHANGELOG.md
# Changelog

### Version 1.0.2 (22 Aug 2021)
- Добавлен вывод времени ожидания между запросами
- Устранена проблема с запуском множества процессов Chrome
- Устранены ошибки

### Version 1.0.1 (18 Aug 2021)
- Изменена ссылка для отслеживания Xbox Series S в магазине Ozon
- Дополнено описание программы в файле [README.md](README.md)

### Version 1.0.0 (17 Aug 2021)
- Релиз программы


test.py
# Программа для тестирования
# Copyright (C) 2021 Evgeniy Ipatov

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


import check_status


links = {
'DNS': 'https://www.dns-shop.ru/product/d0a91343aa2f3332/653-smartfon-xiaomi-redmi-9-64-gb-seryj/',
'Ситилинк': 'https://www.citilink.ru/product/smartfon-xiaomi-redmi-9-64gb-4gb-seryi-3g-4g-6-53-and10-wifi-nfc-gps-1391508/',
'МВидео': 'https://www.mvideo.ru/products/smartfon-xiaomi-redmi-9-464gb-carbon-grey-30050582',
'Эльдорадо': 'https://www.eldorado.ru/cat/detail/smartfon-redmi-9-4-64gb-carbon-grey/',
'Ozon': 'https://www.ozon.ru/product/smartfon-note10-pro-siniy-296727025/',
'Связной': 'https://www.svyaznoy.ru/catalog/phone/224/5787482',
'Сбермегамаркет': 'https://sbermegamarket.ru/catalog/details/smartfon-redmi-9-464gb-carbon-grey-100026886795/',
'Мегафон': 'https://kirov.shop.megafon.ru/mobile/133985.html',
'Gamepark': 'https://www.gamepark.ru/xboxone/games/Cyberpunk2077XboxOne/',
'Technopark': 'https://www.technopark.ru/smartfon-xiaomi-redmi-note-9-64gb-lesnoy-zelenyy/',
'Подпишись.РФ': 'https://подпишись.рф/catalog/Samsung_Galaxy_S21/SM-G991BZVDSER',
'Computeruniverse': 'https://www.computeruniverse.net/ru/p/90825717'
}

for shop, url in links.items():
    print(f'{shop}: {check_status.check_status(shop, url)}')


test/test.py
# Программа для тестирования
# Copyright (C) 2021 Evgeniy Ipatov

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


from ..lib import check_status


links = {
'DNS': 'https://www.dns-shop.ru/product/d0a91343aa2f3332/653-smartfon-xiaomi-redmi-9-64-gb-seryj/',
'Ситилинк': 'https://www.citilink.ru/product/smartfon-xiaomi-redmi-9-64gb-4gb-seryi-3g-4g-6-53-and10-wifi-nfc-gps-1391508/',
'МВидео': 'https://www.mvideo.ru/products/smartfon-xiaomi-redmi-9-464gb-carbon-grey-30050582',
'Эльдорадо': 'https://www.eldorado.ru/cat/detail/smartfon-redmi-9-4-64gb-carbon-grey/',
'Ozon': 'https://www.ozon.ru/product/smartfon-note10-pro-siniy-296727025/',
'Связной': 'https://www.svyaznoy.ru/catalog/phone/224/5787482',
'Сбермегамаркет': 'https://sbermegamarket.ru/catalog/details/smartfon-redmi-9-464gb-carbon-grey-100026886795/',
'Мегафон': 'https://kirov.shop.megafon.ru/mobile/133985.html',
'Gamepark': 'https://www.gamepark.ru/xboxone/games/Cyberpunk2077XboxOne/',
'Technopark': 'https://www.technopark.ru/smartfon-xiaomi-redmi-note-9-64gb-lesnoy-zelenyy/',
'Подпишись.РФ': 'https://подпишись.рф/catalog/Samsung_Galaxy_S21/SM-G991BZVDSER',
'Computeruniverse': 'https://www.computeruniverse.net/ru/p/90825717'
}

for shop, url in links.items():
    print(f'{shop}: {check_status.check_status(shop, url)}')


sound_test.py
# Программа для проверки звукового оповещения
# Copyright (C) 2021 Evgeniy Ipatov

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


import winsound


for index in range(0, 10):
    winsound.Beep(800, 1000)


xbox_hunter.py
# This file is part of XboxHunter project.
# Copyright (C) 2021 Evgeniy Ipatov

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


from utils.check_status import *
from utils.get_links import *
from utils.email_notify import EmailNotity
import random
import time
import sys
import webbrowser
import winsound


# Выбор консоли
search_mode = input('\nВыберите консоль (введите число и нажмите Enter):\n1 - Xbox Series X\n2 - Xbox Series S\n')
while True:
    if search_mode == '1':
        print('\nВыбрана Xbox Series X')
        target_links = get_xsx_links()
        break
    elif search_mode == '2':
        print('\nВыбрана Xbox Series S')
        target_links = get_xss_links()
        break
    else:
        search_mode = input('\nВыберите консоль (введите число и нажмите Enter):\n1 - Xbox Series X\n2 - Xbox Series S\n')

if len(target_links) == 0:
    print('Не найдены ссылки для отслеживания. Выход из программы')
    sys.exit()

# Запрос электронной почты для отправки уведомлений
send_notify = input('\nХотите, чтобы программа отправляла уведомление на электронную почту? Введите Д или Н (также допускается Y или N):\n')
if send_notify.lower() in ['д', 'y']:
    send_notify = True
    notify_object = EmailNotity()
else:
    send_notify = False

# Проверка наличия консоли
exit_flag = False
while True:
    print(f'\n{time.strftime("%H:%M:%S", time.localtime())}')
    for shop, url in target_links.items():
        result = check_status(url)
        if result:
            print(f'{shop:.<20}' + f'{status_message():.>30}')
            lucky_url = url
            exit_flag = True
            break
        else:
            print(f'{shop:.<20}' + f'{status_message():.>30}')
    if exit_flag:
        break
    else:
        delay = random.randint(60, 180)
        print(f'\nСледующий запрос через: {delay} сек.')
        time.sleep(delay)

# Открытие ссылки в браузере
try:
    webbrowser.open_new_tab(lucky_url)
except:
    print('\nНе удалось открыть ссылку в браузере')

# Отправка уведомления на электронную почту
if send_notify:
    notify_object.send_success_message(lucky_url)

# Подача звукового сигнала
try:
    for index in range(0, 100):
        winsound.Beep(800, 1000)
except:
    pass

user_answer = input()


utils/get_links.py
# This file is part of XboxHunter project.
# Copyright (C) 2021 Evgeniy Ipatov

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


import pathlib


def __get_links(folder_name: str) -> dict:
    # Выбор файлов с расширением txt
    p = pathlib.Path('.').parent / folder_name
    p = p.glob('*.txt')
    file_names = sorted([name for name in p if name.is_file()])
    result = {}
    for name in file_names:
        try:
            with open(name, 'rt') as input:
                result[name.stem] = input.readline().replace('\n', '')
        except:
            pass
    return result


def get_xsx_links() -> dict:
    return __get_links('xsx_links')


def get_xss_links() -> dict:
    return __get_links('xss_links')


utils/email_notify.py
# This file is part of XboxHunter project.
# Copyright (C) 2021 Evgeniy Ipatov

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


from getpass import getpass
from email.message import EmailMessage
import smtplib


class EmailNotity():


    def __init__(self) -> None:
        self.__from_email = input('\nВведите адрес электронной почты (поддерживается только gmail), с которой будут отправляться уведомления:\n')
        self.__email_password = getpass('\nВведите пароль электронной почты, указанной выше:\n')
        self.__to_email = input('\nВведите адрес электронной почты, на которую будет отправлено уведомление:\n')
        self.send_start_message()


    def __send_message(self, message_content: str):
        # Создание сообщения
        msg = EmailMessage()
        msg['Subject'] = 'XboxHunter'
        msg['From'] = self.__from_email
        msg['To'] = self.__to_email
        msg.set_content(message_content)
        # Отправка сообщения
        try:
            server = smtplib.SMTP_SSL('smtp.gmail.com')
            server.ehlo()
            server.login(self.__from_email, self.__email_password)
            server.send_message(msg)
            server.close()
            print('\nУведомление отправлено')
        except Exception as error:
            print('\nОшибка отправки уведомления:\n', error)


    def send_start_message(self):
        self.__send_message('Запущена программа XboxHunter. Ожидайте уведомление со ссылкой.')


    def send_success_message(self, lucky_url: str):
        self.__send_message(lucky_url)


utils/check_status.py
# This file is part of XboxHunter project.
# Copyright (C) 2021 Evgeniy Ipatov

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.


from bs4 import BeautifulSoup as bSoup
import certifi
import requests
from requests_html import HTMLSession


__status_message = ''
__session = HTMLSession()


def check_status(target_url: str) -> bool:
    headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.107 Safari/537.36'}
    global __status_message
    __status_message = ''
    try:
        if 'https://xn--d1aiavecq8cxb.xn--p1ai' in target_url:
            global __session
            response = __session.get(target_url, headers=headers, timeout=30, verify=certifi.where())
        else:
            response = requests.get(target_url, headers=headers, timeout=30, verify=certifi.where())
    except:
        __status_message = 'Ошибка запроса'
        return False
    if not response.ok:
        __status_message = 'Отказано в доступе'
        return False
    if 'https://xn--d1aiavecq8cxb.xn--p1ai' in target_url:
        response.html.render(timeout=60)
        parsed_page = bSoup(response.html.html, 'html.parser')
    else:
        parsed_page = bSoup(response.text, 'html.parser')

    if 'mvideo.ru' in target_url:
        result = False

    elif 'eldorado.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('a', class_='gs-avail-btn item_subscription_link no-mobile')]
        result = len(data) == 0

    elif 'dns-shop.ru' in target_url:
        result = False

    elif 'citilink.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('span', class_='Button__text jsButton__text')]
        result = '\n                В корзину\n            ' in data

    elif 'ozon.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('div', class_='kxa6')]
        result = 'Добавить в корзину' in data

    elif 'svyaznoy.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('span', class_='b-main-btn__text')]
        result = 'Ð\x9aÑ\x83Ð¿Ð¸Ñ\x82Ñ\x8c' in data

    elif 'sbermegamarket.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('button', class_='buy-button__button btn sm btn-block')]
        result = 'Ð\x9aÑ\x83Ð¿Ð¸Ñ\x82Ñ\x8c' in data

    elif 'megafon.ru' in target_url:
        result = False

    elif 'gamepark.ru' in target_url:
        data_1 = [element.text for element in parsed_page.find_all('a', class_='subscribe_no_by modal_link')]
        result_1 = not 'Узнать о наличии' in data_1
        data_2 = [element.text for element in parsed_page.find_all('a', class_='book_now is_magazin modal_link')]
        result_2 = 'Только в магазине' in data_2
        result = result_1 or result_2

    elif 'technopark.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('span', class_='buy')]
        result = 'Купить' in data

    # подпишись.рф
    elif 'xn--d1aiavecq8cxb.xn--p1ai' in target_url:
        data = [element.text for element in parsed_page.find_all('button', class_='btn btn-lg btn-rounded btn-order mb-sm-3')]
        result = 'Оформить заявку' in data

    elif 'computeruniverse.net' in target_url:
        result = False

    elif 'kcentr.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('a', class_='product-cart _card jsAddToCart btn _block _red')]
        result = ' Добавить в корзину' in data

    elif '1c-interes.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('a', class_='btn_order product_buy_button btn btn_orange retailRocket-bakset-btn order-link')]
        result = '\nВ корзину\n' in data

    elif 'onlinetrade.ru' in target_url:
        data = [element.text for element in parsed_page.find_all('h3')]
        result = not 'Снят с продажи' in data

    else:
        result = False

    if result:
        __status_message = 'Есть в наличии'
        __session.close()
    else:
        __status_message = 'Нет в наличии'
    
    return result


def status_message() -> str:
    return __status_message
